pkg/manager/tidbngmonitoring/ng_monitoring_manager_test.go

```diff
@@ -0,0 +1,850 @@
+				getSts: func() (*apps.StatefulSet, *apps.StatefulSet) {
+					sts := &apps.StatefulSet{}
+					sts.Name = NGMonitoringName("ngm")
+					sts.Namespace = "default"
+					return sts.DeepCopy(), sts.DeepCopy()
+				},
+				createStatefulSetErr: fmt.Errorf("shouldn't create sts"),
+				updateStatefulSetErr: fmt.Errorf("should update sts"),
+				expectFn: func(tngm *v1alpha1.TidbNGMonitoring, err error) {
+					g.Expect(err).Should(HaveOccurred())
+					g.Expect(err.Error()).Should(ContainSubstring("should update sts"))
+				},
+			},
+		}
+
+		for _, testcase := range cases {
+			t.Logf("testcase: %s", testcase.name)
+
+			deps := controller.NewFakeDependencies()
+			indexer := deps.KubeInformerFactory.Apps().V1().StatefulSets().Informer().GetIndexer()
+			manager := NewNGMonitorManager(deps)
+
+			tngm := &v1alpha1.TidbNGMonitoring{}
+			tngm.Name = "ngm"
+			tngm.Namespace = "default"
+			if testcase.setTNGM != nil {
+				testcase.setTNGM(tngm)
+			}
+
+			// mock old and new sts
+			if testcase.getSts != nil {
+				old, new := testcase.getSts()
+				patch := gomonkey.ApplyFunc(GenerateNGMonitoringStatefulSet, func(tngm *v1alpha1.TidbNGMonitoring, cm *corev1.ConfigMap) (*apps.StatefulSet, error) {
+					return new, nil
+				})
+				defer patch.Reset()
+				if old != nil {
+					indexer.Add(old)
+				}
+			}
+			// mock result of status update
+			patch := gomonkey.ApplyPrivateMethod(reflect.TypeOf(manager), "populateStatus", func(_ *ngMonitoringManager, _ *v1alpha1.TidbNGMonitoring, _ *apps.StatefulSet) error {
+				return testcase.populateStatusErr
+			})
+			defer patch.Reset()
+			// mock result of configmap update
+			patch = gomonkey.ApplyPrivateMethod(reflect.TypeOf(manager), "syncConfigMap", func(_ *ngMonitoringManager, _ *v1alpha1.TidbNGMonitoring, _ *apps.StatefulSet) (*corev1.ConfigMap, error) {
+				return nil, testcase.syncConfigMapErr
+			})
+			defer patch.Reset()
+			// mock result of sts creation
+			deps.StatefulSetControl.(*controller.FakeStatefulSetControl).SetCreateStatefulSetError(testcase.createStatefulSetErr, 0)
+			// mock result of sts update
+			patch = gomonkey.ApplyFunc(mngerutils.UpdateStatefulSet, func(_ controller.StatefulSetControlInterface, _ runtime.Object, _, _ *apps.StatefulSet) error {
+				return testcase.updateStatefulSetErr
+			})
+			defer patch.Reset()
+
+			err := manager.syncCore(tngm)
+			testcase.expectFn(tngm, err)
+		}
+	})
+
+	t.Run("confirmStatefulSetIsUpgrading", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+
+		type testcase struct {
+			name     string
+			setSts   func(*apps.StatefulSet)
+			hasPod   bool
+			setPod   func(*corev1.Pod)
+			expectFn func(bool, error)
+		}
+
+		cases := []testcase{
+			{
+				name: "sts is upgrading",
+				setSts: func(sts *apps.StatefulSet) {
+					sts.Status.CurrentRevision = "v1"
+					sts.Status.UpdateRevision = "v2"
+					sts.Status.ObservedGeneration = 1000
+				},
+				hasPod: false,
+				setPod: nil,
+				expectFn: func(is bool, err error) {
```

Comment:
rename `is` to `upgrading` may be better?

Reply:
done